### ISSUE TITLE
type checking rule results in error where a warning may be more appropriate

### ISSUE THREAD

**Issue by jaymccon** [awslabs/cfn-python-lint#180]
*cfn-lint version: 0.3.2*

*E3012 - "Check resource properties values" is not always an error*

While I agree providing strict type checking (#42) is a good thing to do, I think there are 2 distinct cases. One which will cause a stack to fail, which should be an error, and another where nothing bad would happen, it's just not best practice, this should be a warning.

If there is separate tests for weakly typed values that only generates warnings, this would allow a user to disable this best practice test (if they disagree or don't care) without having to disable type checking all together.

Example (using lifecycle hook's HeartBeatTimeout propoerty, which requires and integer:

Error, will cause stack failure:
```
"MyHook": {
  "Type" : "AWS::AutoScaling::LifecycleHook",
  "Properties" : {
    "HeartbeatTimeout" : false,
    ...
  }
}
```
Warning, will not cause stack failure, just a potential bad practice:
```
"MyHook": {
  "Type" : "AWS::AutoScaling::LifecycleHook",
  "Properties" : {
    "HeartbeatTimeout" : "10",
    ...
  }
}
```

If agreement is reached that this is reasonable/desired I'd be happy to put in a PR.

**Comment by fatbasstard** [comment]
I understand the issue, and it makes sense. But it is pretty much unmaintainable. Currently the rule is pretty simple: If the specification says it's a String, it's a String.

Making a differentiation in "Errors" and "Warnings" is only possible with custom logic. So keep track of which properties actually give errors and which don't. That's a maintenance you don't want to do.

How "annoying" this rule might seem, it's pretty good rule which results in a cleaner and better maintainable codebase. If the dev decided a `HeartbeatTimeout" is a String, let's keep it a String. What if they did it because they also support (or will support) notation like `10 minutes` instead of just 10?

So personally my advice is to keep the rule as in and make the CloudFormation code "better" by working strict, that's a common think in programming languages, why not in CloudFormation. 
And the rule can also be disabled of course 😉 

So in short: Rules can always be improved, but this will end up in a really specific list that needs to be maintained and tested.

**Comment by cmmeyer** [comment]
I agree that this would be challenging to maintain, since the logic is situational and would need to be coded manually.  The base rule should stick to the spec.  If someone can craft more accurate overlays for the special cases, we can always incorporate them and retire this rule later.  

You're going to see the resource spec become more and more the source of truth in CloudFormation, so it's best to treat it literally (even if the APIs don't always line up right now).

**Comment by jaymccon** [comment]
I guess I'm missing some complexities as to why this would be difficult to maintain. 

My current use-case involves linting templates that I am not the author of (so can't enforce best practices) and it is important to know whether these templates are broken (so disabling type checking rule is not desirable). If this is useful to others, check out my fork https://github.com/awslabs/cfn-python-lint/compare/master...jaymccon:feature/weaktyping

**Comment by fatbasstard** [comment]
There's more context in https://github.com/awslabs/cfn-python-lint/issues/42...

> In my experience as long as a property can be resolved into the correct type, CloudFormation does not fail and simply converts the type behind the scenes.

So that is actually not true. It can actually break CloudFormation...

**Comment by jaymccon** [comment]
My mistake has been that all the templates I've tested against have already gone through a functional test, so the cases where incorrect types do cause errors have already been filtered out. 

Apologies for the run around on this one, I'm in agreement that if there's a chance of type mismatch causing an error it's a better approach to be strict.
